src/main/java/org/truffleruby/language/methods/GetMethodObjectNode.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2013, 2020 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.language.methods;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.RootCallTarget;
+import com.oracle.truffle.api.Truffle;
+import com.oracle.truffle.api.TruffleLanguage;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.CachedContext;
+import com.oracle.truffle.api.dsl.GenerateUncached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.Frame;
+import com.oracle.truffle.api.frame.FrameDescriptor;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import org.truffleruby.RubyContext;
+import org.truffleruby.RubyLanguage;
+import org.truffleruby.core.array.ArrayUtils;
+import org.truffleruby.core.cast.BooleanCastNode;
+import org.truffleruby.core.cast.NameToJavaStringNode;
+import org.truffleruby.core.klass.RubyClass;
+import org.truffleruby.core.method.RubyMethod;
+import org.truffleruby.language.RubyBaseNode;
+import org.truffleruby.language.RubyContextSourceNode;
+import org.truffleruby.language.RubyNode;
+import org.truffleruby.language.RubyRootNode;
+import org.truffleruby.language.Visibility;
+import org.truffleruby.language.arguments.RubyArguments;
+import org.truffleruby.language.control.RaiseException;
+import org.truffleruby.language.dispatch.DispatchConfiguration;
+import org.truffleruby.language.dispatch.DispatchNode;
+import org.truffleruby.language.objects.AllocationTracing;
+import org.truffleruby.language.objects.LogicalClassNode;
+import org.truffleruby.language.objects.MetaClassNode;
+
+@GenerateUncached
+public abstract class GetMethodObjectNode extends RubyBaseNode {
+    public static GetMethodObjectNode create() {
+        return GetMethodObjectNodeGen.create();
+    }
+
+    public abstract RubyMethod executeGetMethodObject(Frame frame, Object self, Object name,
+            DispatchConfiguration dispatchConfig, Frame callerFrame);
+
+    protected int getCacheLimit() {
+        return RubyLanguage.getCurrentLanguage().options.METHOD_LOOKUP_CACHE;
+    }
+
+    @Specialization(guards = {
+            "dispatchConfig == cachedDispatchConfig",
+            "self == cachedSelf",
```

Comment:
Why is it caching on `self` here? `cachedSelf` seems unused.

In general, I think we shouldn't do more caching here than in the original node, inline caching already happens in LookupMethodOnSelfNode and that should be enough.

The RubyMethod often does not escape, so the allocation can also be removed in those cases.

Reply:
>Why is it caching on self here? cachedSelf seems unused.

Hmm, I thought we should cache all passed args; and yes, I missed `cachedSelf` in the arguments below (should replace `self`).

What about `"name.equals(cachedName)"`? 

```
=== Found 1 compilation blacklist violations ===

Blacklisted method
  java.lang.Object.equals(Object)
called from
  org.truffleruby.language.methods.GetMethodObjectNodeGen.executeGetMethodObject(GetMethodObjectNodeGen.java:47)
```

Reply:
I think there is no need or major gain from this extra specialization, so could you remove it and only keep `doGeneric` below?
Then getMethodObject() can be inlined for simplicity.

In fact I'd like this node to be as similar as possible to the original one + the DeclarationContext changes to make reviewing easier, right now it's quite difficult because it moved to another file (which is however a good idea).

Reply:
Done: https://github.com/oracle/truffleruby/pull/2177/commits/849d51a6421a841e53c3a01f714f770b17e82d71